app/src/main/kotlin/com/wire/android/ui/calling/incoming/IncomingCallViewModel.kt

```diff
@@ -31,55 +24,52 @@ import javax.inject.Inject
 class IncomingCallViewModel @Inject constructor(
     savedStateHandle: SavedStateHandle,
     private val navigationManager: NavigationManager,
-    private val conversationDetails: ObserveConversationDetailsUseCase,
-    private val allCalls: GetAllCallsUseCase,
+    private val incomingCalls: GetIncomingCallsUseCase,
     private val rejectCall: RejectCallUseCase,
     private val acceptCall: AnswerCallUseCase,
     private val callRinger: CallRinger
 ) : ViewModel() {
-    var callState by mutableStateOf(IncomingCallState())
-        private set
 
     val conversationId: ConversationId = savedStateHandle.get<String>(EXTRA_CONVERSATION_ID)!!.parseIntoQualifiedID()
+    lateinit var observeIncomingCallJob: Job
 
     init {
         viewModelScope.launch {
             callRinger.ring(R.raw.ringing_from_them)
-            launch {
-                conversationDetails(conversationId = conversationId)
-                    .collect { initializeScreenState(conversationDetails = it) }
-            }
-            launch {
+            observeIncomingCallJob = launch {
                 observeIncomingCall()
             }
         }
     }
 
     private suspend fun observeIncomingCall() {
-        allCalls().collect {
-            if (it.first().conversationId == conversationId)
-                when (it.first().status) {
-                    CallStatus.CLOSED -> onCallClosed()
-                    else -> print("DO NOTHING")
-                }
+        incomingCalls().collect { calls ->
+            calls.find { call -> call.conversationId == conversationId }.also {
+                if (it == null)
+                    onCallClosed()
+            }
         }
     }
 
     private fun onCallClosed() {
-        callRinger.stop()
-        viewModelScope.launch { navigationManager.navigateBack() }
+        viewModelScope.launch {
+            navigationManager.navigateBack()
+            callRinger.stop()
+        }
     }
 
     fun declineCall() {
-        callRinger.stop()
         viewModelScope.launch {
+            observeIncomingCallJob.cancel()
             rejectCall(conversationId = conversationId)
+            callRinger.stop()
         }
     }
```

Comment:
shouldn't we also be calling `navigationManager.navigateBack()` here? 🤔  Otherwise it feels like when rejecting the call nothing else will happen (the screen will be kept showing)

Reply:
Good catch 👍